bindings/python/tests/implementations/test_sentencepiece.py

```diff
@@ -14,10 +16,40 @@ def test_train_from_iterator(self):
 
 
 class TestSentencePieceUnigram:
+    def test_train(self, tmpdir):
+        p = tmpdir.mkdir("tmpdir").join("file.txt")
+        p.write("A first sentence\nAnother sentence\nAnd a last one")
+
+        tokenizer = SentencePieceUnigramTokenizer()
+        tokenizer.train(files=str(p), show_progress=False)
+
+        output = tokenizer.encode("A sentence")
+        assert output.tokens == ["▁A", "▁", "s", "en", "t", "en", "c", "e"]
+
+    def test_train_with_unk_token(self, tmpdir):
+        p = tmpdir.mkdir("tmpdir").join("file.txt")
+        p.write("A first sentence\nAnother sentence\nAnd a last one")
+
+        tokenizer = SentencePieceUnigramTokenizer()
+        tokenizer.train(
+            files=str(p), show_progress=False, special_tokens=["<unk>"], unk_token="<unk>"
+        )
+        output = tokenizer.encode("A sentence 🤗")
+        assert output.ids[-1] == 0
```

Comment:
Could you assert the full list of `ids` and `tokens` (like the other test ?)

Reply:
Regarding the `ids` assert, this is what I initially wanted to do but the algorithm doesn't always calculate exactly the same vocabulary: when I run this test several times one id in the id list changes (between `7` and `8`). 

Concretely, the output of `print(output.ids, output.tokens)` is sometimes:
```
[12, 1, 4, 5, 2, 5, 7, 3, 1, 0] ['▁A', '▁', 's', 'en', 't', 'en', 'c', 'e', '▁', '🤗']
```
and other times:
```
[12, 1, 4, 5, 2, 5, 8, 3, 1, 0] ['▁A', '▁', 's', 'en', 't', 'en', 'c', 'e', '▁', '🤗']
```

Would it be okay if I replaced `assert output.ids[-1] == 0` with `assert (output.ids == [12, 1, 4, 5, 2, 5, 7, 3, 1, 0] or output.ids == [12, 1, 4, 5, 2, 5, 8, 3, 1, 0])? :slightly_smiling_face:

Reply:
Regarding the assert on the token list, the following assert fails: 
``` python
assert output.tokens == ["▁A", "▁", "s", "en", "t", "en", "c", "e", "_", "<unk>"]
```
with the error:
```
>       assert output.tokens == ["▁A", "▁", "s", "en", "t", "en", "c", "e", "_", "<unk>"]
E       AssertionError: assert ['▁A', '▁', '...t', 'en', ...] == ['▁A', '▁', '...t', 'en', ...]
E         At index 8 diff: '▁' != '_'
E         Full diff:
E         - ['▁A', '▁', 's', 'en', 't', 'en', 'c', 'e', '_', '<unk>']
E         ?                                              ^    ^^^^^
E         + ['▁A', '▁', 's', 'en', 't', 'en', 'c', 'e', '▁', '🤗']
E         ?                                              ^    ^
```
Is this indeed the expected behavior?

Reply:
So if the tests are not deterministic definitely let's not add them.

However it's suprising that it's not determinstic though..... It shouldn't be ....

▁ != _  
first is a special unicode character (rarely used that's why google chose it)
Second is regular underscore.

Reply:
<unk> instead of 🤗 is indeed expected

Reply:
Thanks a lot for your answer. I did not make any change for the assert on the `ids` and included the change to put the assert on the `tokens`. Can you check to see if this corresponds to what you expected ? :blush:

Reply:
That's better, and OK I understand better why IDs are not deterministic, they essentially have the same score, so no particular reason that ids should be in a particular order.

Reply:
Thanks for your feedback and for investigating the change in ids. It makes sense!